checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,348 @@
+     * <li>Does the annotation depend on any non-final references?
+     * <li>Does the annotation include any method calls? </l>
+     *
+     *     <p>The CheckController contains an option for just the first, an option for the first and
+     *     second, and an option for the second and third. These are the only possible sets of
+     *     invalidations (according to the document that is reproduced as the JavaDoc on this
+     *     class).
+     *
+     *     <p>The CheckController also contains information that is used in deciding which report is
+     *     issued: it contains separate options for its third option (i.e. the second and third
+     *     items in the list) so that different messages can be issued for field reassignment
+     *     invalidation and invalidation because of a call to a non-side-effect free method.
+     */
+    private Result checkAnno(
+            AnnotationMirror anno,
+            String canonicalTargetName,
+            Node n,
+            CheckController checkController) {
+        UpperBoundAnnotatedTypeFactory factory =
+                (UpperBoundAnnotatedTypeFactory) analysis.getTypeFactory();
+        if (!AnnotationUtils.hasElementValue(anno, "value")) {
+            return Result.SUCCESS;
+        }
+        UBQualifier.LessThanLengthOf qual =
+                (UBQualifier.LessThanLengthOf) UBQualifier.createUBQualifier(anno);
+
+        List<String> dependencies = new ArrayList<>();
+        for (String s : qual.getArrays()) {
+            dependencies.add(s);
+        }
+        dependencies.addAll(qual.getOffsetsAsStrings());
+
+        List<String> canonicalDependencies = new ArrayList<>();
+        for (String s : dependencies) {
+            FlowExpressions.Receiver r = null;
+            try {
+                TreePath path = factory.getPath(n.getTree());
+                if (path != null) {
+                    r = factory.getReceiverFromJavaExpressionString(s, path);
+                }
+
+            } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+            }
+            if (r == null) {
+                return Result.SUCCESS;
+            }
+            canonicalDependencies.add(r.toString());
+            while (r instanceof FlowExpressions.FieldAccess) {
+                // I included an exception here for "this", because otherwise the rules don't make sense - any
+                // field, including final ones, would be invalidated, since it's impossible to know if this is
+                // final or not. But this can't be modified by calling other code, so it's fine.
+                if (checkController == CheckController.NONFINAL_REFS_AND_METHOD_CALLS
+                        || checkController
+                                == CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD) {
+                    if (!((FlowExpressions.FieldAccess) r).getReceiver().toString().equals("this")
```

Comment:
This be` instanceOf ThisReference` rather than comparing strings. Or would `FlowExpressions#isUnmodifiableByOtherCode` work instead of `isFinal`?  That method accounts for receivers correctly.